OpenSim/Common/Exception.h

```diff
@@ -88,11 +127,42 @@ class OSIMCOMMON_API Exception  : public std::exception {
     virtual void print(std::ostream &aOut) const;
 #endif
     // override virtual function from std::exception
-    const char* what() const throw() override {return getMessage();}
+    const char* what() const noexcept override;
 
 //=============================================================================
 };  // END CLASS Exception
 
+
+class IndexOutOfRange : public Exception {
+public:
+    IndexOutOfRange(const std::string& file,
+                    size_t line,
+                    const std::string& func,
+                    size_t index, 
+                    size_t min, 
+                    size_t max) :
+        Exception(file, line, func) {
+        std::string msg = "min = " + std::to_string(min);
+        msg += " max = " + std::to_string(max);
+        msg += " index = " + std::to_string(index);
+
+        addMessage(msg);
+    }
+};
+
+class KeyNotFound : public Exception {
+public:
+    KeyNotFound(const std::string& file,
+                size_t line,
+                const std::string& func,
+                const std::string& key) :
+        Exception(file, line, func) {
+        std::string msg = "Key '" + key + "' not found.";
+
+        addMessage(msg);
```

Comment:
I don't think the message -- "call `addMessage()` immediately after the opening brace" is right. Developers should be free to create the message using temporaries if they want to.

I have added some comments to OpenSim::Exception providing a "blueprint" developers can use to write their exception classes that derive from one of the OpenSim exceptions.